### PR TITLE
Add support for self config source

### PR DESCRIPTION
**Pull request by jrhee17**
Motivation:

This pull request adds support for SELF config source types. This feature allows the control plane to delegate the decision of which protocol to use to the client. 
Reference: https://github.com/envoyproxy/envoy/blob/bd18d0fa7790a7e5fe1a95f6ae271ca02614e36c/api/envoy/config/core/v3/config_source.proto#L239

Note that this implementation may not have been completed from envoy side, so we should use this option with caution.
ref: https://github.com/envoyproxy/envoy/issues/13951

Modifications:

- The dependency relationship between `ResourceNode` and `ConfigSource` has changed. This can be a problem because the configSource specified in `ResourceNode#ConfigSource` can be different from the actual `ConfigSource` used for subscribing. Modified so that `ConfigSource` is computed before creating a `ResourceNode`.
- Renamed `BootstrapApiConfigs` to `ConfigSourceMapper` since it better represents the functionality of the class.
- Added `parentConfigSource` to the logic of computing a new configSource.
- Modified so that subscribed `ResourceNode#configSource` is always non-null. The only case where this is `null` is for static clusters (bootstrap clusters)

Result:

- SELF type configSource is now supported

### CODE REVIEW

File: xds/src/main/java/com/linecorp/armeria/xds/ConfigSourceMapper.java [hunk 1]
```diff
@@ -50,35 +50,28 @@ final class BootstrapApiConfigs {
...
+                                 @Nullable ConfigSource configSource, String resourceName) {
         if (configSource != null && configSource.hasApiConfigSource()) {
             return configSource;
         }
+        if (configSource != null && configSource.hasSelf() && parentConfigSource != null) {
+            return parentConfigSource;
+        }
         if (configSource != null && configSource.hasAds() && bootstrapAdsConfig != null) {
             return bootstrapAdsConfig;
         }
```
Comment (minwoox) [inline xds/src/main/java/com/linecorp/armeria/xds/ConfigSourceMapper.java:77]: nit:
```suggestion
        if (configSource != null) {
            if (configSource.hasApiConfigSource()) {
                return configSource;
            }
            if (configSource.hasSelf() && parentConfigSource != null) {
                return parentConfigSource;
            }
            if (configSource.hasAds() && bootstrapAdsConfig != null) {
                return bootstrapAdsConfig;
            }
        }
```

File: xds/src/main/java/com/linecorp/armeria/xds/ConfigSourceMapper.java [hunk 2]
```diff
@@ -91,22 +84,28 @@ ConfigSource cdsConfigSource(ResourceNode<?> node, String resourceName) {
...
         throw new IllegalArgumentException("Cannot find an RDS config source for route: " + resourceName);
     }
 
-    ConfigSource ldsConfigSource(ResourceNode<?> node) {
-        final ConfigSource configSource = node.configSource();
+    ConfigSource ldsConfigSource(@Nullable ConfigSource parentConfigSource,
+                                 @Nullable ConfigSource configSource) {
         if (configSource != null && configSource.hasApiConfigSource()) {
```
Comment (minwoox) [inline xds/src/main/java/com/linecorp/armeria/xds/ConfigSourceMapper.java:103]: ditto

File: xds/src/main/java/com/linecorp/armeria/xds/ConfigSourceMapper.java [hunk 3]
```diff
@@ -91,22 +84,28 @@ ConfigSource cdsConfigSource(ResourceNode<?> node, String resourceName) {
...
         if (bootstrapAdsConfig != null) {
             return bootstrapAdsConfig;
         }
         throw new IllegalArgumentException("Cannot find an RDS config source for route: " + resourceName);
     }
 
-    ConfigSource ldsConfigSource(ResourceNode<?> node) {
-        final ConfigSource configSource = node.configSource();
+    ConfigSource ldsConfigSource(@Nullable ConfigSource parentConfigSource,
+                                 @Nullable ConfigSource configSource) {
```
Comment (minwoox) [inline xds/src/main/java/com/linecorp/armeria/xds/ConfigSourceMapper.java:102]: Should we also accept the `resourceName` to include it in the exception as other methods do?

**Review by minwoox — COMMENTED** [review]
> Modified so that subscribed ResourceNode#configSource is always non-null.

Let's remove `Nullable` in the `ListenerResourceNode` constructor.

**Comment by jrhee17** [comment]
> Let's remove Nullable in the ListenerResourceNode constructor.

Coincidentally, due to the recent changes now `ListenerRoot` can pass a `null` config source
https://github.com/line/armeria/blob/42edbb2940a0447f8df9dd189fe24acfc842335a/xds/src/main/java/com/linecorp/armeria/xds/ListenerRoot.java#L37

**Review by minwoox — APPROVED** [review]
Looks very useful. Thanks! 👍 👍 👍

File: xds/src/main/java/com/linecorp/armeria/xds/ClusterResourceNode.java [hunk 4]
```diff
@@ -77,11 +77,12 @@ public void doOnChanged(ClusterXdsResource resource) {
         } else if (cluster.hasEdsClusterConfig()) {
             final EdsClusterConfig edsClusterConfig = cluster.getEdsClusterConfig();
             final String serviceName = edsClusterConfig.getServiceName();
-            final ConfigSource configSource = edsClusterConfig.getEdsConfig();
+            final String clusterName = !isNullOrEmpty(serviceName) ? serviceName : cluster.getName();
+            final ConfigSource configSource =
+                    configSourceMapper().edsConfigSource(
+                            configSource(), cluster.getEdsClusterConfig().getEdsConfig(), clusterName);
```
[the remark this answers is not in the document]
Comment (jrhee17) [reply xds/src/main/java/com/linecorp/armeria/xds/ClusterResourceNode.java]: Good point 👍    Looks much cleaner

**Comment by ikhoon** [comment]
Still LGTM.

**Pull request merged**